admin/test/test_release.py

```diff
@@ -53,3 +60,604 @@ def test_non_integer_suffix(self):
+                'clusterhq-staging-docs': {
+                    'en/latest/': 'en/0.3.0/',
+                },
+            },
+            s3_buckets={
+                'clusterhq-staging-docs': {
+                    'index.html': '',
+                    'en/index.html': '',
+                    'en/latest/index.html': '',
+                },
+                'clusterhq-dev-docs': {
+                    '0.3.0-444-gf05215b/sub_index.html': '',
+                },
+            })
+        self.publish_docs(aws, '0.3.0-444-gf05215b', '0.3.1',
+                          environment=Environments.STAGING)
+        self.assertEqual(
+            aws.cloudfront_invalidations, [
+                CreateCloudFrontInvalidation(
+                    cname='docs.staging.clusterhq.com',
+                    paths={
+                        'en/latest/',
+                        'en/latest/sub_index.html',
+                        'en/0.3.1/',
+                        'en/0.3.1/sub_index.html',
+                    }),
+            ])
+
+    def test_creates_cloudfront_invalidation_removed_files(self):
+        """
+        Calling :func:`publish_docs` with a release or documentation version
+        creates an invalidation for
+        - en/latest/
+        - en/<doc_version>/
+        each for every path in the old documentation for <doc_version>.
+        """
+        aws = FakeAWS(
+            routing_rules={
+                'clusterhq-staging-docs': {
+                    'en/latest/': 'en/0.3.0/',
+                },
+            },
+            s3_buckets={
+                'clusterhq-staging-docs': {
+                    'index.html': '',
+                    'en/index.html': '',
+                    'en/latest/index.html': '',
+                    'en/0.3.1/index.html': '',
+                    'en/0.3.1/sub/index.html': '',
+                },
+                'clusterhq-dev-docs': {},
+            })
+        self.publish_docs(aws, '0.3.0-444-gf05215b', '0.3.1',
+                          environment=Environments.STAGING)
+        self.assertEqual(
+            aws.cloudfront_invalidations, [
+                CreateCloudFrontInvalidation(
+                    cname='docs.staging.clusterhq.com',
+                    paths={
+                        'en/latest/',
+                        'en/latest/index.html',
+                        'en/latest/sub/',
+                        'en/latest/sub/index.html',
+                        'en/0.3.1/',
+                        'en/0.3.1/index.html',
+                        'en/0.3.1/sub/',
+                        'en/0.3.1/sub/index.html',
+                    }),
+            ])
+
+    def test_creates_cloudfront_invalidation_previous_version(self):
+        """
+        Calling :func:`publish_docs` with a release or documentation version
+        creates an invalidation for
+        - en/latest/
+        - en/<doc_version>/
+        each for every path in the documentation for version that was
```

Comment:
I guess "each" was meant to go before "version". Whatever it is meant to be I can't make sense of this.